src/ruby/puppet-server-lib/puppet/server/master.rb

```diff
@@ -57,6 +60,53 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def getClassInfoForAllEnvironments()
+    # The clear_environment_settings(env) ensures that the calls which happen
+    # later on to enumerate all of the manifests for each environment are
+    # using the latest environment.conf settings for each environment.  Without
+    # this call, cached (and, therefore, possibly stale) environment.conf
+    # settings would be used.  We want to return the latest data for any calls
+    # made to this method.
+    @env_loader.list.each do |env|
+      Puppet.settings.clear_environment_settings(env)
+    end
+
+    environments =
+      Hash[@env_loader.list.collect do |env|
+       [env.name, self.class.getManifests(env)]
+      end]
```

Comment:
@camlow325 can you please paste that last big comment with your requirements / goals into a new PUP ticket and assign to @hlindberg ?  And maybe create a placeholder ticket for SERVER to link to it?  After that I'm fine w/you merging this (pending review/consensus by whomever you would like to have review it, I haven't really reviewed it thoroughly and probably won't unless you ask me to) and moving on as long as we have the other ticket flagged as a release blocker.

Also FYI I believe that @hlindberg is on PTO for another week or so.  You could check w/Kylo on that.